### PR TITLE
Replace remaining cuda::device operations with native CUDA calls.

### PR DESCRIPTION
**Pull request by waredjeb**
#### PR description:
This PR is part of #386. It replaces the usage of ```cuda::device::count()```, ```cuda::device::get()```, ```cuda::device::set()``` and ```cuda::device::current::get()``` with native cuda calls. 

As in #404 ```HeterogeneousCore/Product``` and ```HeterogeneousCore/Producer``` have not been modified since they will be removed.

#### PR validation:

Code compiles, unit test runs.

### CODE REVIEW

File: HeterogeneousCore/CUDAUtilities/interface/cudaDeviceCount.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#ifndef HeterogenousCore_CUDAUtilities_cudaDeviceCount_h
+#define HeterogenousCore_CUDAUtilities_cudaDeviceCount_h
+
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+
+#include <cuda_runtime.h>
+
+namespace cudautils {
+  inline int cudaDeviceCount() {
```
[the remark this answers is not in the document]
Comment (waredjeb) [reply HeterogeneousCore/CUDAUtilities/interface/cudaDeviceCount.h]: Yes make sense to drop ```cuda```. I'll wait for new comments!

File: SimTracker/TrackerHitAssociation/plugins/ClusterTPAssociationHeterogeneous.cc [hunk 2]
```diff
@@ -191,7 +192,7 @@ void ClusterTPAssociationHeterogeneous::acquireGPUCuda(const edm::HeterogeneousE
   edm::Handle<CUDAProduct<TrackingRecHit2DCUDA>> gh;
   iEvent.getByToken(tGpuHits, gh);
   // temporary check (until the migration)
-  assert(gd->device() == cuda::device::current::get().id());
+  assert(gd->device() == cudautils::currentDevice());
```
Comment (VinInn) [inline SimTracker/TrackerHitAssociation/plugins/ClusterTPAssociationHeterogeneous.cc]: this is obsoleted by #409 .
I think it is easier for everybody not to apply this change here

**Comment by VinInn** [comment]
I think one should also remove all
```#include <cuda/api_wrappers.h>```
and
the corresponding use from the BuildFiles

**Comment by makortel** [comment]
@waredjeb Thanks. Can you confirm that with the last commit the API wrappers are not used anywhere except in `HeterogeneousCore/{Producer,Product}` packages?

**Comment by fwyzard** [comment]
:construction: Validation running at fu-c2a02-35-01:/data/user/fwyzard/patatrack/validation/run.OKjWaeIbaA ...

**Comment by waredjeb** [comment]
> @waredjeb Thanks. Can you confirm that with the last commit the API wrappers are not used anywhere except in `HeterogeneousCore/{Producer,Product}` packages?

Well, not at all. [DataFormats/Math/test/cudaMathTest.cu](https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_0_X_Patatrack/DataFormats/Math/test/cudaMathTest.cu), [DataFormats/Math/test/cudaAtan2Test.cu
 ](https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_0_X_Patatrack/DataFormats/Math/test/cudaAtan2Test.cu), [HeterogeneousCore/CUDACore/interface/GPUCuda.h](https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_0_X_Patatrack/HeterogeneousCore/CUDACore/interface/GPUCuda.h),  [HeterogeneousCore/CUDACore/interface/CUDAScopedContext.h](https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_0_X_Patatrack/HeterogeneousCore/CUDACore/interface/CUDAScopedContext.h), [HeterogeneousCore/CUDACore/interface/CUDAESProduct.h](https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_0_X_Patatrack/HeterogeneousCore/CUDACore/interface/CUDAESProduct.h), and some files in CUDAUtilities still have the API Wrappers. 

(HeterogeneousCore/CUDAUtilities/interface/host_noncached_unique_ptr.h
HeterogeneousCore/CUDAUtilities/interface/launch.h
HeterogeneousCore/CUDAUtilities/src/allocate_host.cc
HeterogeneousCore/CUDAUtilities/src/allocate_device.cc
HeterogeneousCore/CUDAUtilities/BuildFile.xml)
I have errors when I try to remove the api-wrapper.

**Comment by makortel** [comment]
@waredjeb Thanks for the list. I'd leave dealing with these to a follow-up PR(s) of this PR. The remaining cases seem to be either soon-obsolete, or `cuda::throw_if_error()` (more details below).

> [HeterogeneousCore/CUDACore/interface/GPUCuda.h](https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_0_X_Patatrack/HeterogeneousCore/CUDACore/interface/GPUCuda.h)

This file will be removed once #409 gets merged.

> [HeterogeneousCore/CUDACore/interface/CUDAScopedContext.h](https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_0_X_Patatrack/HeterogeneousCore/CUDACore/interface/CUDAScopedContext.h)

The only use seems to be
https://github.com/cms-patatrack/cmssw/blob/614ee0b4e0d1d457d6bbf7d51e0799f330fa242b/HeterogeneousCore/CUDACore/src/CUDAScopedContext.cc#L83
to be replaced with `cudaCheck()` (as discussed [before](https://github.com/cms-patatrack/cmssw/pull/398#issuecomment-548428981), a variant taking a message argument would be nice).

> [HeterogeneousCore/CUDACore/interface/CUDAESProduct.h](https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_0_X_Patatrack/HeterogeneousCore/CUDACore/interface/CUDAESProduct.h)

Also here (after this PR) the only call is
https://github.com/cms-patatrack/cmssw/blob/614ee0b4e0d1d457d6bbf7d51e0799f330fa242b/HeterogeneousCore/CUDACore/interface/CUDAESProduct.h#L62
to be replace with `cudaCheck()`.

> and some files in CUDAUtilities still have the API Wrappers.
> 
> (HeterogeneousCore/CUDAUtilities/interface/host_noncached_unique_ptr.h

`cuda::throw_if_error()` -> `cudaCheck()`

> HeterogeneousCore/CUDAUtilities/interface/launch.h

I don't see API wrappers being used in this file.

> HeterogeneousCore/CUDAUtilities/src/allocate_host.cc
> HeterogeneousCore/CUDAUtilities/src/allocate_device.cc

`cuda::throw_if_error()` -> `cudaCheck()` in both

**Comment by fwyzard** [comment]
No impact on physics or throughput (as expected).

File: DataFormats/Math/test/CholeskyInvert_t.cu [hunk 3]
```diff
@@ -1,5 +1,4 @@
 // nvcc -O3 CholeskyDecomp_t.cu -Icuda-api-wrappers/src/ --expt-relaxed-constexpr -gencode arch=compute_61,code=sm_61 --compiler-options="-Ofast -march=native"
-// add -DDOPROF to run  nvprof --metrics all
```
Comment (fwyzard) [inline DataFormats/Math/test/CholeskyInvert_t.cu]: why do you remove this line ?

File: HeterogeneousCore/CUDAUtilities/interface/cudaDeviceCount.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (fwyzard) [reply HeterogeneousCore/CUDAUtilities/interface/cudaDeviceCount.h]: Agreed that `deviceCount()` would be fine.
Can you make a follow up PR ?

File: HeterogeneousCore/CUDAUtilities/src/allocate_host.cc [hunk 4]
```diff
@@ -3,8 +3,7 @@
 #include "FWCore/Utilities/interface/Likely.h"
 
 #include "getCachingHostAllocator.h"
-
-#include <cuda/api_wrappers.h>
+#include "cuda/api_wrappers.h"
```
Comment (fwyzard) [inline HeterogeneousCore/CUDAUtilities/src/allocate_host.cc]: If still needed, `<cuda/api_wrappers.h>` is more correct than `"cuda/api_wrappers.h"`.

File: HeterogeneousCore/CUDAServices/src/CUDAService.cc [hunk 5]
```diff
@@ -391,6 +392,6 @@ int CUDAService::deviceWithMostFreeMemory() const {
   return device;
 }
 
-void CUDAService::setCurrentDevice(int device) const { cuda::device::current::set(device); }
+void CUDAService::setCurrentDevice(int device) const { cudaCheck(cudaSetDevice(device)); }
 
-int CUDAService::getCurrentDevice() const { return cuda::device::current::get().id(); }
+int CUDAService::getCurrentDevice() const { return cudautils::currentDevice(); }
```
[the remark this answers is not in the document]
Comment (fwyzard) [reply HeterogeneousCore/CUDAServices/src/CUDAService.cc]: agreed - could you create an issue so we don't forget ?

File: DataFormats/Math/test/CholeskyInvert_t.cu (hunk 3 above)
Comment (waredjeb) [reply DataFormats/Math/test/CholeskyInvert_t.cu]: It wasn't my intention, thanks for the fix.

File: HeterogeneousCore/CUDAServices/src/CUDAService.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (makortel) [reply HeterogeneousCore/CUDAServices/src/CUDAService.cc]: > agreed - could you create an issue so we don't forget ?

Done #415.

**Pull request merged**